File: app/temp-swift-backend/src/main/java/at/htlleonding/maturaballmanager/AuthResource.java
package at.htlleonding.maturaballmanager;

import at.htlleonding.maturaballmanager.contracts.oidc.OidcClient;
import at.htlleonding.maturaballmanager.model.keycloak.TokenResponse;
import jakarta.inject.Inject;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import org.eclipse.microprofile.rest.client.inject.RestClient;

import java.util.Map;
import java.util.logging.Logger;

@Path("/auth")
public class AuthResource {

    private static final Logger LOGGER = Logger.getLogger(AuthResource.class.getName());

    @Inject
    @RestClient
    OidcClient oidcClient;

    @POST
    @Path("/login")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Response login(Map<String, String> credentials) {
        String username = credentials.get("username");
        String password = credentials.get("password");

        LOGGER.info("Attempting to get token for username: " + username);

        try {
            TokenResponse tokenResponse = oidcClient.getToken(
                    "password",
                    "maturaballmanager",
                    "C4bc62f5JmmJVMK1Iu6qTmHUdMm0eO8C",
                    username,
                    password,
                    "openid"
            );
            LOGGER.info("Token response: " + tokenResponse);
            return Response.ok(tokenResponse).build();
        } catch (Exception e) {
            LOGGER.severe("Error: " + e.getMessage());
            return Response.status(Response.Status.UNAUTHORIZED).entity("Invalid credentials").build();
        }
    }

    @GET
    @Path("/userinfo")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getUserInfo(@HeaderParam("Authorization") String authorization) {
        if (authorization == null || !authorization.startsWith("Bearer ")) {
            return Response.status(Response.Status.UNAUTHORIZED).entity("Missing or invalid Authorization header").build();
        }

        String token = authorization.substring("Bearer".length()).trim();

        try {
            Map<String, Object> userInfo = oidcClient.getUserInfo(token);
            return Response.ok(userInfo).build();
        } catch (Exception e) {
            LOGGER.severe("Error fetching user info: " + e.getMessage());
            return Response.status(Response.Status.UNAUTHORIZED).entity("Failed to fetch user info").build();
        }
    }
}

File: app/backend/src/main/kotlin/at/htlleonding/maturaballmanager/resources/AuthResource.kt
package at.htlleonding.maturaballmanager.resources

import at.htlleonding.maturaballmanager.services.AuthService
import io.smallrye.mutiny.Uni
import jakarta.inject.Inject
import jakarta.json.JsonString
import jakarta.ws.rs.GET
import jakarta.ws.rs.POST
import jakarta.ws.rs.Path
import jakarta.ws.rs.core.Response
import org.eclipse.microprofile.config.inject.ConfigProperty
import org.eclipse.microprofile.jwt.JsonWebToken

@Path("/auth")
class AuthResource @Inject constructor(private val authService: AuthService) {

    @Inject
    lateinit var jwt: JsonWebToken

    @ConfigProperty(name = "quarkus.oidc.client-id")
    lateinit var clientId: String

    @POST
    @Path("/login")
    fun login(credentials: LoginRequest): Uni<Response> {
        return if (credentials.username.isNullOrBlank() || credentials.password.isNullOrBlank()) {
            Uni.createFrom().item(Response.status(Response.Status.BAD_REQUEST).entity("Invalid credentials").build())
        } else {
            authService.authenticate(credentials.username!!, credentials.password!!)
                .onItem().transform { tokenResponse ->
                    if (tokenResponse != null) {
                        val accessToken = tokenResponse.access_token
                        val refreshToken = tokenResponse.refresh_token

                        Response.ok()
                            .header("Authorization", "Bearer $accessToken")
                            .entity(mapOf("access_token" to accessToken, "refresh_token" to refreshToken))
                            .build()
                    } else {
                        Response.status(Response.Status.UNAUTHORIZED).entity("Authentication failed").build()
                    }
                }
        }
    }

    @POST
    @Path("/refresh")
    fun refreshToken(request: RefreshTokenRequest): Uni<Response> {
        return authService.refreshToken(request.refreshToken.toString())
            .onItem().transform { tokenResponse ->
                if (tokenResponse != null) {
                    Response.ok(tokenResponse).build()
                } else {
                    Response.status(Response.Status.UNAUTHORIZED).entity("Invalid refresh token").build()
                }
            }
    }

    @GET
    @Path("/validate")
    fun validate(): Response {
        return if (jwt.expirationTime > System.currentTimeMillis() / 1000) {
            Response.ok().entity(true).build()
        } else {
            Response.status(Response.Status.UNAUTHORIZED).entity("Token is invalid").build()
        }
    }

    @POST
    @Path("/hasRoles")
    fun hasRoles(roles: List<String>): Response {
        return if (roles.any { getClientRoles().contains(it) }) {
            Response.ok().entity(mapOf("hasRoles" to true)).build()
        } else if(jwt.groups.isEmpty() && roles.isEmpty()) {
            Response.ok().entity(mapOf("hasRoles" to true)).build()
        } else {
            Response.status(Response.Status.FORBIDDEN).entity("User does not have required roles").build()
        }
    }

    private fun getClientRoles(): List<String> {
        val resourceAccess = jwt.getClaim<Map<String, Any>>("resource_access")
        val clientData = resourceAccess?.get(clientId) as? Map<String, Any> ?: return emptyList()

        val rolesObj = clientData["roles"]
        if (rolesObj is List<*>) {
            return rolesObj.mapNotNull {
                if (it is JsonString) {
                    it.string
                } else it as? String
            }
        }
        return emptyList()
    }
}

data class LoginRequest(var username: String? = null, var password: String? = null)
data class RefreshTokenRequest(var refreshToken: String? = null)